core/src/main/java/hudson/slaves/NodeProvisioner.java

```diff
@@ -249,7 +246,29 @@ public void run() {
                                 LOGGER.log(Level.SEVERE, "Unexpected uncaught exception encountered while " 
                                         + "processing provisioned slave " + f.displayName, e);
                             } finally {
-                                itr.remove();
+                                while (true) {
+                                    List<PlannedNode> orig = pendingLaunches.get();
+                                    List<PlannedNode> repl = new ArrayList<PlannedNode>(orig);
+                                    // the contract for List.remove(o) is that the first element i where
+                                    // (o==null ? get(i)==null : o.equals(get(i)))
+                                    // is true will be removed from the list
+                                    // since PlannedNode.equals(o) is not final and we cannot assume
+                                    // that subclasses do not override with an equals which does not
+                                    // assure object identity comparison, we need to manually
+                                    // do the removal based on instance identity not equality
+                                    boolean changed = false;
+                                    for (Iterator<PlannedNode> iterator = repl.iterator(); iterator.hasNext(); ) {
+                                        PlannedNode p = iterator.next();
+                                        if (p == f) {
+                                            iterator.remove();
+                                            changed = true;
+                                            break;
+                                        }
+                                    }
+                                    if (!changed || pendingLaunches.compareAndSet(orig, repl)) {
+                                        break;
```

Comment:
I think you are analyzing the call flow incorrectly here. This whole change is just updating the list, not calling the listener. we are in a `while (true)` to ensure the lock-free update. if we were not changed then the node has already been removed from the list somehow - should never happen as we are the only thread removing elements from this list (trust but verify) - and we do not need to replace the original list with the new list... similarly if there was a change and we replaced the original (when we captured it) with our replacement then we are done and we can escape the `while (true)` to proceed with the call to `f.spent()`.

The only thread with the responsibility for calling `f.spent()` is the current thread, so we will only ever call it once. I could replace the `!changed` with `assert changed` but I do not see much value since we know there are copy&paste fans out there and we want to move to more lock-free approaches so better to illustrate how to handle it